File: python/src/program/envelope.py
"""Order envelope builders for the Lightcone SDK.

Provides fluent builder pattern for constructing limit and trigger orders.
Matches TS program/envelope.ts.
"""

from typing import Optional

from solders.keypair import Keypair
from solders.pubkey import Pubkey

from .types import SignedOrder, OrderSide
from .orders import sign_order
from ..shared.types import Side, TimeInForce, TriggerType, SubmitTriggerOrderRequest
from ..shared.scaling import OrderbookDecimals, scale_price_size


class LimitOrderEnvelope:
    """Fluent builder for limit orders, matching TS LimitOrderEnvelope."""

    def __init__(self):
        self._nonce: int = 0
        self._maker: Optional[Pubkey] = None
        self._market: Optional[Pubkey] = None
        self._base_mint: Optional[Pubkey] = None
        self._quote_mint: Optional[Pubkey] = None
        self._side: OrderSide = OrderSide.BID
        self._amount_in: int = 0
        self._amount_out: int = 0
        self._expiration: int = 0

    def nonce(self, nonce: int) -> "LimitOrderEnvelope":
        self._nonce = nonce
        return self

    def maker(self, maker: Pubkey) -> "LimitOrderEnvelope":
        self._maker = maker
        return self

    def market(self, market: Pubkey) -> "LimitOrderEnvelope":
        self._market = market
        return self

    def base_mint(self, mint: Pubkey) -> "LimitOrderEnvelope":
        self._base_mint = mint
        return self

    def quote_mint(self, mint: Pubkey) -> "LimitOrderEnvelope":
        self._quote_mint = mint
        return self

    def bid(self) -> "LimitOrderEnvelope":
        self._side = OrderSide.BID
        return self

    def ask(self) -> "LimitOrderEnvelope":
        self._side = OrderSide.ASK
        return self

    def side(self, side: Side) -> "LimitOrderEnvelope":
        self._side = OrderSide(int(side))
        return self

    def amount_in(self, amount: int) -> "LimitOrderEnvelope":
        self._amount_in = amount
        return self

    def amount_out(self, amount: int) -> "LimitOrderEnvelope":
        self._amount_out = amount
        return self

    def expiration(self, expiration: int) -> "LimitOrderEnvelope":
        self._expiration = expiration
        return self

    def apply_scaling(self, price: str, size: str, decimals: OrderbookDecimals) -> "LimitOrderEnvelope":
        """Apply price/size scaling to set amount_in and amount_out."""
        scaled = scale_price_size(price, size, int(self._side), decimals)
        self._amount_in = scaled.amount_in
        self._amount_out = scaled.amount_out
        return self

    def finalize(self) -> SignedOrder:
        """Build the unsigned SignedOrder."""
        assert self._maker is not None, "maker is required"
        assert self._market is not None, "market is required"
        assert self._base_mint is not None, "base_mint is required"
        assert self._quote_mint is not None, "quote_mint is required"

        return SignedOrder(
            nonce=self._nonce,
            maker=self._maker,
            market=self._market,
            base_mint=self._base_mint,
            quote_mint=self._quote_mint,
            side=self._side,
            amount_in=self._amount_in,
            amount_out=self._amount_out,
            expiration=self._expiration,
        )

    def sign(self, keypair: Keypair) -> SignedOrder:
        """Build and sign the order."""
        order = self.finalize()
        sign_order(order, keypair)
        return order


class TriggerOrderEnvelope:
    """Fluent builder for trigger orders, matching TS TriggerOrderEnvelope."""

    def __init__(self):
        self._limit = LimitOrderEnvelope()
        self._trigger_price: str = "0"
        self._trigger_type: TriggerType = TriggerType.STOP_LOSS
        self._time_in_force: TimeInForce = TimeInForce.GTC

    def nonce(self, nonce: int) -> "TriggerOrderEnvelope":
        self._limit.nonce(nonce)
        return self

    def maker(self, maker: Pubkey) -> "TriggerOrderEnvelope":
        self._limit.maker(maker)
        return self

    def market(self, market: Pubkey) -> "TriggerOrderEnvelope":
        self._limit.market(market)
        return self

    def base_mint(self, mint: Pubkey) -> "TriggerOrderEnvelope":
        self._limit.base_mint(mint)
        return self

    def quote_mint(self, mint: Pubkey) -> "TriggerOrderEnvelope":
        self._limit.quote_mint(mint)
        return self

    def bid(self) -> "TriggerOrderEnvelope":
        self._limit.bid()
        return self

    def ask(self) -> "TriggerOrderEnvelope":
        self._limit.ask()
        return self

    def side(self, side: Side) -> "TriggerOrderEnvelope":
        self._limit.side(side)
        return self

    def amount_in(self, amount: int) -> "TriggerOrderEnvelope":
        self._limit.amount_in(amount)
        return self

    def amount_out(self, amount: int) -> "TriggerOrderEnvelope":
        self._limit.amount_out(amount)
        return self

    def expiration(self, expiration: int) -> "TriggerOrderEnvelope":
        self._limit.expiration(expiration)
        return self

    def apply_scaling(self, price: str, size: str, decimals: OrderbookDecimals) -> "TriggerOrderEnvelope":
        self._limit.apply_scaling(price, size, decimals)
        return self

    def trigger_price(self, price: str) -> "TriggerOrderEnvelope":
        self._trigger_price = price
        return self

    def trigger_type(self, tt: TriggerType) -> "TriggerOrderEnvelope":
        self._trigger_type = tt
        return self

    def stop_loss(self) -> "TriggerOrderEnvelope":
        self._trigger_type = TriggerType.STOP_LOSS
        return self

    def take_profit(self) -> "TriggerOrderEnvelope":
        self._trigger_type = TriggerType.TAKE_PROFIT
        return self

    def time_in_force(self, tif: TimeInForce) -> "TriggerOrderEnvelope":
        self._time_in_force = tif
        return self

    def gtc(self) -> "TriggerOrderEnvelope":
        self._time_in_force = TimeInForce.GTC
        return self

    def ioc(self) -> "TriggerOrderEnvelope":
        self._time_in_force = TimeInForce.IOC
        return self

    def fok(self) -> "TriggerOrderEnvelope":
        self._time_in_force = TimeInForce.FOK
        return self

    def sign(self, keypair: Keypair) -> SignedOrder:
        """Build and sign the underlying order."""
        return self._limit.sign(keypair)

    def to_submit_trigger_request(self, order: SignedOrder, orderbook_id: str) -> SubmitTriggerOrderRequest:
        """Convert to a SubmitTriggerOrderRequest."""
        from .orders import signature_hex
        return SubmitTriggerOrderRequest(
            maker=str(order.maker),
            nonce=order.nonce,
            market_pubkey=str(order.market),
            base_token=str(order.base_mint),
            quote_token=str(order.quote_mint),
            side=int(order.side),
            amount_in=order.amount_in,
            amount_out=order.amount_out,
            expiration=order.expiration,
            signature=signature_hex(order),
            orderbook_id=orderbook_id,
            trigger_price=self._trigger_price,
            trigger_type=int(self._trigger_type),
            time_in_force=int(self._time_in_force),
        )

File: typescript/src/program/envelope.ts
import { Keypair, PublicKey } from "@solana/web3.js";
import bs58 from "bs58";
import type { ClientContext } from "../context";
import { requireSigningStrategy } from "../context";
import { SdkError } from "../error";
import { RetryPolicy } from "../http";
import {
  privyOrderFromLimitEnvelope,
  privyOrderFromTriggerEnvelope,
} from "../privy";
import {
  scalePriceSize,
  validateRawAmounts,
  validateSignedFields,
  validateTriggerPrice,
  type OrderbookRules,
} from "../shared/scaling";
import { isUserCancellation } from "../shared/signing";
import { Orderbooks, type OrderBookPair } from "../domain/orderbook";
import type {
  DepositSource,
  SubmitOrderRequest,
  TimeInForce,
  TriggerType,
} from "../shared";
import type { SubmitOrderResponse } from "../domain/order/client";
import type { TriggerOrderResponse } from "../domain/order/client";
import { ProgramSdkError } from "./error";
import {
  generateSalt,
  hashOrderHex,
  signOrder,
  signOrderFull,
  toSubmitRequest,
  type SubmitRequestOptions,
} from "./orders";
import { OrderSide, type SignedOrder } from "./types";

interface OrderFields {
  nonce?: number;
  salt?: bigint;
  maker?: PublicKey;
  market?: PublicKey;
  baseMint?: PublicKey;
  quoteMint?: PublicKey;
  side?: OrderSide;
  amountIn?: bigint;
  amountOut?: bigint;
  expiration: bigint;
  priceRaw?: string;
  sizeRaw?: string;
  depositSource?: DepositSource;
}

function defaultFields(): OrderFields {
  return {
    expiration: 0n,
  };
}

function toUnsignedOrder(fields: OrderFields): Omit<SignedOrder, "signature"> {
  if (fields.nonce === undefined) throw ProgramSdkError.missingField("nonce");
  if (!fields.maker) throw ProgramSdkError.missingField("maker");
  if (!fields.market) throw ProgramSdkError.missingField("market");
  if (!fields.baseMint) throw ProgramSdkError.missingField("base_mint");
  if (!fields.quoteMint) throw ProgramSdkError.missingField("quote_mint");
  if (fields.side === undefined) throw ProgramSdkError.missingField("side");
  if (fields.amountIn === undefined) throw ProgramSdkError.missingField("amount_in");
  if (fields.amountOut === undefined) throw ProgramSdkError.missingField("amount_out");

  const salt = fields.salt ?? generateSalt();
  fields.salt = salt;
  validateSignedFields(fields.amountIn, fields.amountOut, salt, fields.nonce);
  return {
    nonce: fields.nonce,
    salt,
    maker: fields.maker,
    market: fields.market,
    baseMint: fields.baseMint,
    quoteMint: fields.quoteMint,
    side: fields.side,
    amountIn: fields.amountIn,
    amountOut: fields.amountOut,
    expiration: fields.expiration,
  };
}

export interface OrderEnvelope {
  nonce(value: number): this;
  salt(value: bigint): this;
  maker(value: PublicKey): this;
  market(value: PublicKey): this;
  baseMint(value: PublicKey): this;
  quoteMint(value: PublicKey): this;
  bid(): this;
  ask(): this;
  side(value: OrderSide): this;
  amountIn(value: bigint): this;
  amountOut(value: bigint): this;
  expiration(value: bigint): this;
  price(value: string): this;
  size(value: string): this;
  depositSource(value: DepositSource): this;
  payload(): Omit<SignedOrder, "signature">;
  sign(keypair: Keypair, orderbook: OrderBookPair, rules: OrderbookRules): SubmitOrderRequest;
  finalize(signatureBase58: string, orderbook: OrderBookPair, rules: OrderbookRules): SubmitOrderRequest;
}

class BaseEnvelope {
  protected readonly fields: OrderFields;

  constructor(fields?: OrderFields) {
    this.fields = fields ? { ...fields } : defaultFields();
  }

  nonce(value: number): this {
    this.fields.nonce = value;
    return this;
  }

  salt(value: bigint): this {
    this.fields.salt = value;
    return this;
  }

  maker(value: PublicKey): this {
    this.fields.maker = value;
    return this;
  }

  market(value: PublicKey): this {
    this.fields.market = value;
    return this;
  }

  baseMint(value: PublicKey): this {
    this.fields.baseMint = value;
    return this;
  }

  quoteMint(value: PublicKey): this {
    this.fields.quoteMint = value;
    return this;
  }

  bid(): this {
    this.fields.side = OrderSide.BID;
    return this;
  }

  ask(): this {
    this.fields.side = OrderSide.ASK;
    return this;
  }

  side(value: OrderSide): this {
    this.fields.side = value;
    return this;
  }

  amountIn(value: bigint): this {
    this.fields.amountIn = value;
    return this;
  }

  amountOut(value: bigint): this {
    this.fields.amountOut = value;
    return this;
  }

  expiration(value: bigint): this {
    this.fields.expiration = value;
    return this;
  }

  price(value: string): this {
    this.fields.priceRaw = value;
    return this;
  }

  size(value: string): this {
    this.fields.sizeRaw = value;
    return this;
  }

  depositSource(value: DepositSource): this {
    this.fields.depositSource = value;
    return this;
  }

  payload(): Omit<SignedOrder, "signature"> {
    return toUnsignedOrder(this.fields);
  }

  getNonce(): number | undefined {
    return this.fields.nonce;
  }

  getSalt(): bigint | undefined {
    return this.fields.salt;
  }

  getMaker(): PublicKey | undefined {
    return this.fields.maker;
  }

  getMarket(): PublicKey | undefined {
    return this.fields.market;
  }

  getBaseMint(): PublicKey | undefined {
    return this.fields.baseMint;
  }

  getQuoteMint(): PublicKey | undefined {
    return this.fields.quoteMint;
  }

  getSide(): OrderSide | undefined {
    return this.fields.side;
  }

  getAmountIn(): bigint | undefined {
    return this.fields.amountIn;
  }

  getAmountOut(): bigint | undefined {
    return this.fields.amountOut;
  }

  getExpiration(): bigint {
    return this.fields.expiration;
  }

  getDepositSource(): DepositSource | undefined {
    return this.fields.depositSource;
  }

  /**
   * Auto-fill market, base_mint, quote_mint, and salt from the orderbook
   * if not explicitly set by the caller.
   */
  protected autoFillFromOrderbook(orderbook: OrderBookPair): void {
    if (!this.fields.market) {
      this.fields.market = new PublicKey(orderbook.marketPubkey);
    }
    if (this.fields.salt === undefined) {
      this.fields.salt = generateSalt();
    }
    if (!this.fields.baseMint) {
      this.fields.baseMint = new PublicKey(orderbook.base.pubkey);
    }
    if (!this.fields.quoteMint) {
      this.fields.quoteMint = new PublicKey(orderbook.quote.pubkey);
    }
  }

  /** Construct or preflight the signed ratio using fetched immutable rules. */
  protected applyRules(rules: OrderbookRules): void {
    if (this.fields.side === undefined) throw ProgramSdkError.missingField("side");
    if (this.fields.amountIn !== undefined && this.fields.amountOut !== undefined) {
      validateRawAmounts(this.fields.amountIn, this.fields.amountOut, this.fields.side, rules);
      return;
    }
    if (this.fields.amountIn !== undefined || this.fields.amountOut !== undefined) {
      throw ProgramSdkError.missingField("amount_in and amount_out must be supplied together");
    }
    if (!this.fields.priceRaw) throw ProgramSdkError.missingField("price");
    if (!this.fields.sizeRaw) throw ProgramSdkError.missingField("size");
    const scaled = scalePriceSize(this.fields.priceRaw, this.fields.sizeRaw, this.fields.side, rules);
    this.fields.amountIn = scaled.amountIn;
    this.fields.amountOut = scaled.amountOut;
  }

  protected finalizeWithHexSignature(
    signatureHex: string,
    orderbookId: string,
    options: SubmitRequestOptions = {}
  ): SubmitOrderRequest {
    const unsigned = toUnsignedOrder(this.fields);
    return toSubmitRequest(
      {
        ...unsigned,
        signature: Buffer.from(signatureHex, "hex"),
      },
      orderbookId,
      {
        ...options,
        depositSource: this.fields.depositSource,
      }
    );
  }
}

export class LimitOrderEnvelope extends BaseEnvelope implements OrderEnvelope {
  private timeInForceValue?: TimeInForce;

  static new(): LimitOrderEnvelope {
    return new LimitOrderEnvelope();
  }

  timeInForce(value: TimeInForce): this {
    this.timeInForceValue = value;
    return this;
  }

  getTimeInForce(): TimeInForce | undefined {
    return this.timeInForceValue;
  }

  sign(keypair: Keypair, orderbook: OrderBookPair, rules: OrderbookRules): SubmitOrderRequest {
    this.autoFillFromOrderbook(orderbook);
    this.applyRules(rules);
    const signed = signOrderFull(this.payload(), keypair, rules);
    return toSubmitRequest(signed, orderbook.orderbookId, {
      timeInForce: this.timeInForceValue,
      depositSource: this.getDepositSource(),
    });
  }

  finalize(signatureBase58: string, orderbook: OrderBookPair, rules: OrderbookRules): SubmitOrderRequest {
    this.autoFillFromOrderbook(orderbook);
    this.applyRules(rules);
    const signatureHex = Buffer.from(bs58.decode(signatureBase58)).toString("hex");
    return this.finalizeWithHexSignature(signatureHex, orderbook.orderbookId, {
      timeInForce: this.timeInForceValue,
    });
  }

  async submit(
    client: ClientContext,
    orderbook: OrderBookPair
  ): Promise<SubmitOrderResponse> {
    const rules = await new Orderbooks(client).decimals(orderbook.orderbookId);
    const strategy = requireSigningStrategy(client);
    this.autoFillFromOrderbook(orderbook);
    this.applyRules(rules);

    // Nonce cache: cache if explicitly set, auto-populate from cache if not
    if (this.fields.nonce !== undefined) {
      client.setOrderNonce?.(this.fields.nonce);
    } else {
      this.fields.nonce = client.orderNonce?.() ?? 0;
    }

    switch (strategy.type) {
      case "native": {
        const request = this.sign(strategy.keypair, orderbook, rules);
        const url = `${client.http.baseUrl()}/api/orders/submit`;
        return client.http.post<SubmitOrderResponse, SubmitOrderRequest>(
          url,
          request,
          RetryPolicy.None
        );
      }
      case "walletAdapter": {
        const unsigned = this.payload();
        const hash = hashOrderHex({ ...unsigned, signature: Buffer.alloc(64) });
        const sigBytes = await strategy.signer
          .signMessage(new TextEncoder().encode(hash))
          .catch((err: unknown) => {
            const msg = err instanceof Error ? err.message : String(err);
            if (isUserCancellation(msg)) throw SdkError.userCancelled();
            throw SdkError.signing(msg);
          });
        const request = this.finalize(bs58.encode(sigBytes), orderbook, rules);
        const url = `${client.http.baseUrl()}/api/orders/submit`;
        return client.http.post<SubmitOrderResponse, SubmitOrderRequest>(
          url,
          request,
          RetryPolicy.None
        );
      }
      case "privy": {
        const envelope = privyOrderFromLimitEnvelope(this, orderbook.orderbookId);
        const url = `${client.http.baseUrl()}/api/privy/sign_and_send_order`;
        return client.http.post(url, { wallet_id: strategy.walletId, order: envelope }, RetryPolicy.None);
      }
    }
  }
}

export class TriggerOrderEnvelope extends BaseEnvelope implements OrderEnvelope {
  private timeInForceValue?: TimeInForce;
  private triggerPriceValue?: string;
  private triggerTypeValue?: TriggerType;

  static new(): TriggerOrderEnvelope {
    return new TriggerOrderEnvelope();
  }

  timeInForce(value: TimeInForce): this {
    this.timeInForceValue = value;
    return this;
  }

  triggerPrice(value: string): this {
    this.triggerPriceValue = value;
    return this;
  }

  triggerType(value: TriggerType): this {
    this.triggerTypeValue = value;
    return this;
  }

  takeProfit(price: string): this {
    this.triggerPriceValue = price;
    this.triggerTypeValue = "TP" as TriggerType;
    return this;
  }

  stopLoss(price: string): this {
    this.triggerPriceValue = price;
    this.triggerTypeValue = "SL" as TriggerType;
    return this;
  }

  gtc(): this {
    this.timeInForceValue = "GTC" as TimeInForce;
    return this;
  }

  ioc(): this {
    this.timeInForceValue = "IOC" as TimeInForce;
    return this;
  }

  fok(): this {
    this.timeInForceValue = "FOK" as TimeInForce;
    return this;
  }

  alo(): this {
    this.timeInForceValue = "ALO" as TimeInForce;
    return this;
  }

  getTimeInForce(): TimeInForce | undefined {
    return this.timeInForceValue;
  }

  getTriggerPrice(): string | undefined {
    return this.triggerPriceValue;
  }

  getTriggerType(): TriggerType | undefined {
    return this.triggerTypeValue;
  }

  sign(keypair: Keypair, orderbook: OrderBookPair, rules: OrderbookRules): SubmitOrderRequest {
    const trigger = this.requireTriggerFields();
    validateTriggerPrice(trigger.price, rules.priceDecimals);
    this.autoFillFromOrderbook(orderbook);
    this.applyRules(rules);
    const signed = signOrderFull(this.payload(), keypair, rules);
    return toSubmitRequest(signed, orderbook.orderbookId, {
      timeInForce: this.timeInForceValue,
      triggerPrice: Number(trigger.price),
      triggerType: trigger.type,
      depositSource: this.getDepositSource(),
    });
  }

  finalize(signatureBase58: string, orderbook: OrderBookPair, rules: OrderbookRules): SubmitOrderRequest {
    const trigger = this.requireTriggerFields();
    validateTriggerPrice(trigger.price, rules.priceDecimals);
    this.autoFillFromOrderbook(orderbook);
    this.applyRules(rules);
    const signatureHex = Buffer.from(bs58.decode(signatureBase58)).toString("hex");
    return this.finalizeWithHexSignature(signatureHex, orderbook.orderbookId, {
      timeInForce: this.timeInForceValue,
      triggerPrice: Number(trigger.price),
      triggerType: trigger.type,
    });
  }

  async submit(
    client: ClientContext,
    orderbook: OrderBookPair
  ): Promise<TriggerOrderResponse> {
    const rules = await new Orderbooks(client).decimals(orderbook.orderbookId);
    const strategy = requireSigningStrategy(client);
    const trigger = this.requireTriggerFields();
    validateTriggerPrice(trigger.price, rules.priceDecimals);
    this.autoFillFromOrderbook(orderbook);
    this.applyRules(rules);

    // Nonce cache: cache if explicitly set, auto-populate from cache if not
    if (this.fields.nonce !== undefined) {
      client.setOrderNonce?.(this.fields.nonce);
    } else {
      this.fields.nonce = client.orderNonce?.() ?? 0;
    }

    switch (strategy.type) {
      case "native": {
        const request = this.sign(strategy.keypair, orderbook, rules);
        const url = `${client.http.baseUrl()}/api/orders/submit`;
        return client.http.post<TriggerOrderResponse, SubmitOrderRequest>(
          url,
          request,
          RetryPolicy.None
        );
      }
      case "walletAdapter": {
        const unsigned = this.payload();
        const hash = hashOrderHex({ ...unsigned, signature: Buffer.alloc(64) });
        const sigBytes = await strategy.signer
          .signMessage(new TextEncoder().encode(hash))
          .catch((err: unknown) => {
            const msg = err instanceof Error ? err.message : String(err);
            if (isUserCancellation(msg)) throw SdkError.userCancelled();
            throw SdkError.signing(msg);
          });
        const request = this.finalize(bs58.encode(sigBytes), orderbook, rules);
        const url = `${client.http.baseUrl()}/api/orders/submit`;
        return client.http.post<TriggerOrderResponse, SubmitOrderRequest>(
          url,
          request,
          RetryPolicy.None
        );
      }
      case "privy": {
        const envelope = privyOrderFromTriggerEnvelope(this, orderbook.orderbookId);
        const url = `${client.http.baseUrl()}/api/privy/sign_and_send_order`;
        return client.http.post(url, { wallet_id: strategy.walletId, order: envelope }, RetryPolicy.None);
      }
    }
  }

  private requireTriggerFields(): { price: string; type: TriggerType } {
    if (this.triggerPriceValue === undefined) {
      throw ProgramSdkError.missingField("trigger_price");
    }
    if (!this.triggerTypeValue) {
      throw ProgramSdkError.missingField("trigger_type");
    }
    return { price: this.triggerPriceValue, type: this.triggerTypeValue };
  }
}

export function signPayload(
  payload: Omit<SignedOrder, "signature">,
  keypair: Keypair,
  rules: OrderbookRules
): string {
  return signOrder(
    { ...payload, signature: Buffer.alloc(64) },
    keypair,
    rules
  ).toString("hex");
}
